File: src/features/typing/data/ts/path_sum_iii.ts
class TreeNode {
    val: number;
    left: TreeNode | null;
    right: TreeNode | null;
    constructor(val = 0, left: TreeNode | null = null, right: TreeNode | null = null) {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

function pathSum(root: TreeNode | null, targetSum: number): number {
    const count = new Map<number, number>([[0, 1]]);
    const dfs = (node: TreeNode | null, presum: number): number => {
        if (node === null) {
            return 0;
        }
        presum += node.val;
        let total = count.get(presum - targetSum) ?? 0;
        count.set(presum, (count.get(presum) ?? 0) + 1);
        total += dfs(node.left, presum) + dfs(node.right, presum);
        count.set(presum, count.get(presum)! - 1);
        return total;
    };
    return dfs(root, 0);
}

File: src/features/typing/data/dart/path_sum_iii.dart
class Solution {
  int pathSum(TreeNode? root, int targetSum) {
    final count = <int, int>{0: 1};
    return _dfs(root, 0, targetSum, count);
  }

  int _dfs(TreeNode? node, int presum, int targetSum, Map<int, int> count) {
    if (node == null) {
      return 0;
    }
    presum += node.val;
    var total = count[presum - targetSum] ?? 0;
    count[presum] = (count[presum] ?? 0) + 1;
    total += _dfs(node.left, presum, targetSum, count);
    total += _dfs(node.right, presum, targetSum, count);
    count[presum] = count[presum]! - 1;
    return total;
  }
}
